CommandTree Prototype

figpope opened the pull request:
@ms-lolo, I'm opening this PR so we can brainstorm / refine this new structure based on the design we thought up last Thursday. I've converted all the commands in `rats-devtools`, but left the old versions so you can get a quick feel for the difference in the structure

Tab-completion isn't working in my Codespaces env on either your branch or this one, which likely is related to the `/home/vscode/.zshrc:105: command not found: register-python-argcomplete` error I get when opening a new shell in a component

This adds a `CommandTree` object that can be converted into click commands (and can be easily adapted to handle other frameworks as well), along with a `CommandServiceTree` that can be converted lazily if necessary

It re-implements most of the `argparse` specific functionality of [`simple-parsing`](https://github.com/lebrice/SimpleParsing), but using `click` and introduces an intermediate representation + pattern that allows generalization to other frameworks

It's a mostly unopinionated structure that encourages users to follow CLI UX best practices, while abstracting away the complexity of the underlying framework

Things have been designed to allow power users to directly interact with the underlying framework to avoid the abstraction becoming a limitation

figpope commented:
> @microsoft-github-policy-service agree

figpope commented:
@microsoft-github-policy-service agree

ms-lolo commented:
i think i shuffled some stuff around in my branch at 2am so could you help me understand some of the ideas here in relation to those? we can have the thread in this pr or mine, which ever you prefer.

maybe we can split the work in 3 to discuss them individually and talk about their relationships?

`rats.annotations` has the decorator logic that is not related to services
`rats.cli` adds uses of annotations specific to cli commands, and if you're happy with `click`, i would just couple `rats.cli` to that library for now.
`rats.devtools` for the concrete combination of `rats.apps/annotations/cli` to define the devtools cli experience (but not any concrete commands)

can you help me understand the relationship with your ideas to these three areas? are there other abstractions?

figpope commented:
Let's sync here since you've just kicked off the convo in this thread :)

Just saw your updates (and updated this branch to match), and def agree with splitting things into those three categories:

* `rats.annotations` - I think this is effectively a standalone library, and I just added tests to all of the code you added in #157 so I could be confident in my changes / I usually use a flow closer to TDD. I like the way this works, and the way you're handling annotations w/ services + namespaces + groups makes it easy to separate (and is similar to how I factored the infra stuff when rewriting it with Apps & Services)
* `rats.cli` - Most of these changes are related to the DX around linking commands together. I think we took somewhat similar approaches, but with different usability trade-offs in the pre-2:30am implementation.
    * The new approach you're using with the `CommandContainer`  and `cli.command()` is similar to what I did with the `.subcommands("<command prefix>"` structure
    * We took opposite approaches, w.r.t. coupling to the underlying libraries, though. I split things into a structure for composition, and then a conversion step to go from the abstract CLI command tree to the concrete `click` implementation. It looks like you added abstractions to make it as easy as possible to compose `click.Command`s and `click.Group`s, and added structure after they're constructed to make lazy-loading easier
    * The trade-offs we're looking at seem to be:
        * **Tightly couple to one library** vs **Add abstraction to support many libraries** - I def agree with your point about `click` owning most of the market share, but there are a lot of nascent libraries in the space, too ([awesome-cli-frameworks#python](https://github.com/shadawck/awesome-cli-frameworks?tab=readme-ov-file#python)). It shouldn't be much effort to maintain a couple of different reference implementations (like we discussed w/ argparse + click), and structuring things this way means that users of the framework can write + maintain their own bespoke converters
        * **Make it as easy as possible to write a complete CLI in one file** vs **Make it easy to decompose the commands you're writing into individual testable functions** - When I wrote the `CommandTree` stuff in a single file, it composes pretty simply, since you end up with a single declarative structure for your whole nested CLI. Our approaches are pretty compatible here if we look at the `CommandContainer` as a way to generate `click.Groups` (or arbitrary `CommandTrees`). Since I used a recursive tree traversal to convert things into `click` constructs, we can mix and match approaches pretty easily, providing an on-ramp from easy to write to easy to maintain
        * **Directly expose the CLI-related details** vs **Abstract away the CLI-related details** - I think this one is slightly related to the first trade-off, but has some DX related elements. I tried to approach the problem from the perspective of "someone who knows Python, but doesn't know any CLI frameworks and wants something more batteries-included-but-optional than argparse", so all users need to know to use the `CommandTree` abstraction is how to write Python functions and dataclasses, as opposed to investing in learning a complete / more opinionated framework. We're both targeting the opposite audiences here than in the bullet above this one, and I think this split effectively lies along the "new user" vs "power user" spectrum
    * `rats.devtools` - 100% agree here, and was following your example structure to flesh out how the two patterns interact

The easiest way (from my POV) to sync on design is to look at how commands are defined in the two frameworks:

**>rats-devtools/src/python/rats/docs/_commands.py**
```python
    @apps.service(PluginServices.command("build-tutorial-notebooks"))
    def build_tutorial_notebooks_command(self) -> click.Command:
        @click.command()
        def build_tutorial_notebooks() -> None:
            """
            Build the tutorial notebooks section of each component's documentation.

            Converts each *.py jupytext file in the component's docs/_tutorial_notebook_sources
            folder into a markdown notebook in the component's docs/_tutorial_notebooks folder.
            """
            _build_tutorial_notebooks()

        return build_tutorial_notebooks
```

**>rats-devtools/src/python/rats/docs/_v2.py**
```python
    @cli.command(cli.CommandId("foocli2"))
    @click.argument("example")
    def baz(self, example: str) -> None:
        print(f"doing baz with {example}")
```
    
**>rats-devtools/src/python/rats/docs/_commands.py (w/ CommandTree)**
```python
    @apps.group(command_tree.CommandTreeServices.GROUPS.subcommands("rats-devtools docs"))
    def build_tutorial_notebooks_command_tree(self) -> command_tree.CommandTree:
        return command_tree.CommandTree(
            name="build-tutorial-notebooks",
            description="Build the tutorial notebooks section of each component's documentation.",
            handler=_build_tutorial_notebooks,
        )
```

**example of single file CommandTree w/ grouped args**
```python
import logging
from dataclasses import dataclass

from rats import command_tree


@dataclass(frozen=True)
class ParentArgs:
    """Arguments for the parent command."""

    a: int
    """The first number to operate on."""

    b: str
    """The second number to operate on."""


def build_tutorial_notebooks_command_tree() -> command_tree.CommandTree:
    return command_tree.CommandTree(
        name="math-cli",
        description="It does simple math",
        kwargs_class=ParentArgs,
        handler=lambda: logging.basicConfig(level=logging.INFO),
        children=(
            command_tree.CommandTree(
                name="add",
                description="Add two numbers",
                handler=lambda parent_args: print(parent_args.a - int(parent_args.b)),
            ),
            command_tree.CommandTree(
                name="subtract",
                description="Subtract two numbers",
                handler=lambda parent_args: print(parent_args.a - int(parent_args.b)),
            ),
            command_tree.CommandTree(
                name="multiply",
                description="Multiply two numbers",
                handler=multiply,
            ),
            command_tree.CommandTree(
                name="divide",
                description="Divide two numbers",
                kwargs_class=DivideArgs,
                handler=divide,
            ),
        ),
    )


def multiply(parent_args: ParentArgs) -> None:
    print(parent_args.a * int(parent_args.b))


@dataclass(frozen=True)
class DivideArgs:
    """Arguments for the divide command."""

    throw_error: bool = False
    """Whether to throw an error if the dividing by zero."""


def divide(parent_args: ParentArgs, divide_args: DivideArgs) -> None:
    if parent_args.b == "0":
        if divide_args.throw_error:
            raise ValueError("Cannot divide by zero!")
        else:
            print("Warning: Dividing by zero!")
    else:
        print(parent_args.a / int(parent_args.b))


if __name__ == "__main__":
    example_command_tree = build_tutorial_notebooks_command_tree()
    command_tree.CommandTreeClickExecutable(example_command_tree).execute()
```

ms-lolo commented on rats-devtools/src/python/rats/annotations/_functions.py:
```diff
@@ -60,11 +67,11 @@ def make(self, name: str) -> AnnotationsContainer:
-def annotation(namespace: str, group_id: NamedTuple) -> AnnotationDecorator:
-    def decorator(fn: Callable[..., Any]) -> Callable[..., Any]:
+def annotation(namespace: str, group_id: NamedTuple) -> Callable[[DecoratorType], DecoratorType]:
+    def decorator(fn: DecoratorType) -> DecoratorType:
```
ugh ty, this stuff was a nightmare and the reason i ended up ignoring the linting and pushing at 2am after fighting it all night

figpope replied:
Happy to help! Type hints on decorators are def tricky, especially when you have functions acting as factories for them!

ms-lolo commented on rats-devtools/src/python/rats/docs/_commands.py:
```diff
@@ -20,34 +24,70 @@ class RatsCiCommands(apps.AnnotatedContainer):
     def __init__(self, app: apps.Container) -> None:
         self._app = app
 
+    @apps.group(command_tree.CommandTreeServices.GROUPS.subcommands("rats-devtools docs"))
```
with my plugin classes, i think i've separated the command from where it gets assigned to a parent command. is that possible here? can we avoid mentioning the "rats-devtools" and "docs" commands?

figpope replied:
That's 100% possible! I just chose the `rats-devtools docs` as the "subcommands" group because it makes it easy to see where the command fits in with everything else when looking at a single function

You could replace it with an arbitrary value, as long as it's a group ServiceId that subcommands are registered to

I tend to lean towards the parent command getting to decide whether to expose a subcommand group ServiceId (open for extension) or explicitly list the subcommands (closed for extension). What about you?

ms-lolo replied:
let me tinker in my branch tonight and form some opinions :) i'm going to convert some of my commands to different styles like you did here and see how it compares

ms-lolo commented on rats-devtools/src/python/rats/docs/_commands.py:
```diff
@@ -58,171 +98,182 @@ def build_tutorial_notebooks() -> None:
-        args = [
-            "--config-file",
-            str(mkdocs_staging_config),
+def _do_mkdocs_things(cmd: str) -> None:
```
i'd like to keep these in the class so we can slowly move the side effects out to other services

figpope replied:
Fixed in [e6e0781](https://github.com/microsoft/rats/pull/161/commits/e6e07815cdc3c0d9b7dd0c9cc30fe917c4b2259f)

ms-lolo commented:
i like all these changes and i want to get them merged into my branch quickly before i break any of your tests. if you switch my side back to namedtuples, i think we're good to merge without having anything break in my branch but i would also strongly suggest moving your tree commands out to another file with its own container and adding your container to the plugin in the package. it will just give us a better view to discuss differences in the dx and keep tinkering without too much worry and merging often. left a few other comments on changes but most can be resolved in other prs!

figpope commented:
> i like all these changes and i want to get them merged into my branch quickly before i break any of your tests. if you switch my side back to namedtuples, i think we're good to merge without having anything break in my branch but i would also strongly suggest moving your tree commands out to another file with its own container and adding your container to the plugin in the package. it will just give us a better view to discuss differences in the dx and keep tinkering without too much worry and merging often. left a few other comments on changes but most can be resolved in other prs!

I split these up in [5281d11](https://github.com/microsoft/rats/pull/161/commits/5281d11eeead85b0734177766eb0c14c54194e73), undid the NamedTuple -> Dataclass switch, and switched the default base command back to using the CommandContainer root command (so the behavior in your PR is what the rats-devtools command will execute)

AFAIK, this should be ready for merge!
